proposals/0569-multiline-strings.rst

```diff
@@ -0,0 +1,629 @@
+
+This proposal is heavily influenced by Java's text blocks in `JEP 378 <https://openjdk.org/jeps/378>`_.
+
+Motivation
+----------
+
+Currently you could do this with ``unlines`` (more commonly) or line continuations (`"gaps" <https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-200002.6>`_):
+
+::
+
+  s1 = unlines
+    [ "line 1"
+    , "line 2"
+    , "line 3"
+    ]
+
+  s1_2 = "line 1\n\
+         \line 2\n\
+         \line 3\n"
+
+Using ``unlines`` is decent, but it's a bit verbose. The verbosity becomes more apparent when building multiline strings with double quotes, e.g. rendering Haskell code or SQL queries.
+
+::
+
+  s2 = unlines
+    [ "x :: String"
+    , "x = a ++ b"
+    , "  where"
+    , "    a = \"Hello \""
+    , "    b = \"world!\""
+    ]
+
+  s3 = unlines
+    [ "SELECT \"id\", \"name\""
+    , "FROM \"user\""
+    , "WHERE \"name\" ILIKE 'Alice%'"
+    ]
+
+Compare with multiline strings:
+
+::
+
+  s1' =
+    """
+    line 1
+    line 2
+    line 3
+    """
+
+  s2' =
+    """
+    x :: String
+    x = a ++ b
+      where
+        a = "Hello "
+        b = "world!"
+    """
+
+  s3' =
+    """
+    SELECT "id", "name"
+    FROM "user"
+    WHERE "name" ILIKE 'Alice%'
+    """
+
+Third party libraries also provide this functionality with quasiquoters, e.g. ``heredoc`` or libraries that also do interpolation like ``string-interpolate``. But Template Haskell is not great:
+
+* It makes compilation difficult in certain environments (e.g. GHCJS)
+* It can slow down compilation
+* Some people avoid it as much as possible out of principle
+* It's a rather heavyweight tool for a seemingly lightweight syntactic feature
+
+Proposed Change Specification
+-----------------------------
+
+A working prototype is available at `brandonchinn178/string-syntax <https://github.com/brandonchinn178/string-syntax>`_.
+
+#. Lex ``"""`` as an `additional string delimiter <https://gitlab.haskell.org/ghc/ghc/-/blob/8c0ea25fb4a27d4729aabf73f4c00b912bb0c58d/compiler/GHC/Parser/Lexer.x#L577>`_ when the ``MultilineStrings`` language extension is enabled
+
+#. Post-process the string in the following steps:
+
+   #. Collapse line continuations
+
+   #. Remove common whitespace prefix in every line
+
+      * Ignore any characters preceding the first newline
+      * Blank lines and lines with only whitespace should not be included in this calculation
+
+   #. Remove exactly one newline from the beginning of the string (if one exists)
+
+   #. Interpret escape sequences (occurs after removing whitespace prefix so that literal ``\n`` characters are not included)
+
+#. After parsing, it becomes indistinguishable to the equivalent single-quoted string (modulo annotations for exact-printing)
+
+Line terminators will match whatever line terminators the user is using in the file.
+
+Examples
+--------
+
+Line continuations
```

Comment:
Can we not support line continuations in `"""` strings? I believe they only exist in `"` strings to enable a more noisy version of multi-line strings. Mixing the two seems like unnecessary complexity.

Reply:
Line continuations could still be useful if you don't want a trailing newline but want to keep the closing `"""` delimiter on the next line (see the "Trailing newline" example)

> I believe they only exist in `"` strings to enable a more noisy version of multi-line strings.

I don't know if that's necessarily true. What if someone wants 200 characters in a single line of output, but they don't want the string in the source code to have a line length of 200?

> Mixing the two seems like unnecessary complexity.

I agree it's confusing to read as a user, and I wouldn't recommend doing so, but I don't know if it needs to be forbidden by the syntax. IMO everything you can do with normal strings, you should be able to do with multiline strings; any deviations in behavior from normal strings should have a strong motivation. I don't see line continuations as adding that much complexity to this syntax: banning line continuations just saves one post processing step.